File: android/src/main/kotlin/me/yohom/amapbase/map/model/UnifiedRouteOverlay.kt
package me.yohom.amapbase.map.model

import com.amap.api.maps.AMap
import com.amap.api.maps.model.LatLng
import me.yohom.amapbase.map.overlay.DrivingRouteOverlay

internal class UnifiedRouteOverlay(
        val type: String,
        val from: LatLng,
        val to: LatLng,
        private val passby: List<LatLng>,
        private val drivePath: UnifiedDrivePath
) {
    fun applyTo(map: AMap) {
        DrivingRouteOverlay(map, from, to, passby, drivePath).apply {
            removeFromMap()
            addToMap()
            zoomToSpan()
        }
    }
}

class UnifiedDrivePath(
        val strategy: String,
        val tolls: Double,
        val tollDistance: Double,
        val totalTrafficlights: Int,
        val steps: List<UnifiedDriveStep>,
        val restriction: Int
)

class UnifiedDriveStep(
        val instruction: String,
        val orientation: String,
        val road: String,
        val distance: Double,
        val tolls: Double,
        val tollDistance: Double,
        val tollRoad: String,
        val duration: Double,
        val polyline: List<LatLng>,
        val action: String,
        val assistantAction: String,
        val routeSearchCityList: List<UnifiedRouteSearchCity>,
        val TMCs: List<UnifiedTMC>
)

class UnifiedRouteSearchCity(val districts: List<UnifiedDistrict>)

class UnifiedTMC(
        val distance: Int,
        val status: String,
        val polyline: List<LatLng>)

class UnifiedDistrict(
        val districtName: String,
        val districtAdcode: String
)

File: ios/Classes/map/model/UnifiedRouteOverlay.m
//
// Created by Yohom Bao on 2018-12-12.
//

#import "UnifiedRouteOverlay.h"
#import "UnifiedAMapOptions.h"
#import "MJExtension.h"
#import "AMapCommonObj.h"


@implementation UnifiedRouteOverlay {

}
+ (instancetype)initWithJson:(NSString *)json {
    [UnifiedRouteOverlay mj_setupObjectClassInArray:^NSDictionary * {
        return @{@"passby": @"LatLng"};
    }];
    [UnifiedDrivePath mj_setupObjectClassInArray:^NSDictionary * {
        return @{@"steps": @"UnifiedDriveStep"};
    }];
    [UnifiedDriveStep mj_setupObjectClassInArray:^NSDictionary * {
        return @{
                @"routeSearchCityList": @"UnifiedRouteSearchCity",
                @"TMCs": @"UnifiedTMC",
                @"polyline": @"LatLng",
        };
    }];
    [UnifiedRouteSearchCity mj_setupObjectClassInArray:^NSDictionary * {
        return @{@"districts": @"UnifiedDistrict"};
    }];
    [UnifiedTMC mj_setupObjectClassInArray:^NSDictionary * {
        return @{@"polyline": @"LatLng"};
    }];
    return [UnifiedRouteOverlay mj_objectWithKeyValues:json];
}

@end

@implementation UnifiedDrivePath {

}


@end

@implementation UnifiedDriveStep {

}

@end

@implementation UnifiedRouteSearchCity {

}

@end

@implementation UnifiedTMC {

}

@end

@implementation UnifiedDistrict {

}
@end